### PR TITLE
fix: make stepk match galsim

### PR DESCRIPTION
**Pull request by beckermr**
This PR gets stepk for interpolated images to match galsim exactly. 🎉 

xref: #189

### CODE REVIEW

File: jax_galsim/interpolatedimage.py [hunk 1]
```diff
@@ -1175,7 +1175,7 @@ def _flux_frac(a, x, y, cenx, ceny):
     dy = jnp.reshape(dy, (a.shape[0], a.shape[1], 1))
     d = jnp.arange(a.shape[0])
     d = jnp.reshape(d, (1, 1, -1))
-    msk = (jnp.abs(dx) <= d) & (jnp.abs(dx) <= d)
+    msk = (jnp.abs(dx) <= d) & (jnp.abs(dy) <= d)
```
Comment (beckermr) [inline jax_galsim/interpolatedimage.py:1166]: This was a legit bug. :(

File: jax_galsim/interpolatedimage.py [hunk 2]
```diff
@@ -687,13 +686,15 @@ def _maxk(self):
...
+            # conservative when computing maxk
+            return self._getMaxK(self._jax_aux_data["calculate_maxk"]) * 1.1
 
     @property
     def _stepk(self):
         if self._jax_aux_data["_force_stepk"]:
-            _, minor = compute_major_minor_from_jacobian(self._jac_arr.reshape((2, 2)))
-            return self._jax_aux_data["_force_stepk"] * minor
+            major, _ = compute_major_minor_from_jacobian(self._jac_arr.reshape((2, 2)))
+            return self._jax_aux_data["_force_stepk"] * major
```
Comment (beckermr) [inline jax_galsim/interpolatedimage.py]: This was also a bug. 🤦

File: jax_galsim/interpolatedimage.py [hunk 3]
```diff
@@ -1193,28 +1192,20 @@ def _calculate_size_containing_flux(image, thresh):
     cenx, ceny = image.center.x, image.center.y
     x, y = image.get_pixel_centers()
     fluxes = _flux_frac(image.array, x, y, cenx, ceny)
-    msk = fluxes >= -jnp.inf
-    fluxes = jnp.where(msk, fluxes, jnp.max(fluxes))
+    # msk = fluxes >= -jnp.inf
+    # fluxes = jnp.where(msk, fluxes, jnp.max(fluxes))
     d = jnp.arange(image.array.shape[0]) + 1.0
```
Comment (ismael-mendoza) [inline jax_galsim/interpolatedimage.py]: curious if this assumes the image is a square? GalSim does something with the minimum in the x and y direction to set this maximum distance no?

Comment (beckermr) [reply jax_galsim/interpolatedimage.py]: Yes we can compute a min across the axis sizes.

**Comment by beckermr** [comment]
Test failures here are due to maxk not matching.

File: jax_galsim/interpolatedimage.py (hunk 2 above)
Comment (ismael-mendoza) [reply jax_galsim/interpolatedimage.py]: could you explain if you don't mind? I couldn't find the logic in GalSim multiplying times the jacobian major/minor anywhere

Comment (beckermr) [reply jax_galsim/interpolatedimage.py]: It is in the transform class and this bit of code undoes what the transform class will do later.

**Comment by beckermr** [comment]
Let's go ahead with this PR for now. I still want to look into maxk more, but this PR leaves the code in a pretty good working state.

**Comment by beckermr** [comment]
@ismael-mendoza this is ready for review!

**Comment by ismael-mendoza** [comment]
thanks! I can take a look today

File: jax_galsim/interpolatedimage.py
Comment (ismael-mendoza) [inline jax_galsim/interpolatedimage.py]: not sure whether to leave this for another PR... but do we need `self._original._wcs._maxScale()` here like for stepk we have `self._original._wcs._minScale()`?

And the Transform `super()` there applies `minor`, not `major`...

I see that GalSim does it here: https://github.com/GalSim-developers/GalSim/blob/ee177aa50424dbbc5cec4a71cadd6297109c2aa1/galsim/interpolatedimage.py#L626, but I don't think we do?

Comment (ismael-mendoza) [inline jax_galsim/interpolatedimage.py]: this might depend on the changes from the above comment, but do we need `minor` in this line to remove the `minor` from the `Transform` above?

File: jax_galsim/interpolatedimage.py [hunk 4]
```diff
@@ -1193,28 +1191,18 @@ def _calculate_size_containing_flux(image, thresh):
...
-    # below we use a linear interpolation table to find the maximum size
-    # in pixels that contains a given flux (called thresh here)
-    # expfac controls how much we oversample the interpolation table
-    # in order to return a more accurate result
-    # we have it hard coded at 4 to compromise between speed and accuracy
-    expfac = 4.0
-    dint = jnp.arange(image.array.shape[0] * expfac) / expfac + 1.0
-    fluxes = jnp.interp(dint, d, fluxes)
-    msk = fluxes <= thresh
+    d = jnp.arange(min(image.array.shape[0], image.array.shape[1])) + 1.0
```
Comment (ismael-mendoza) [inline jax_galsim/interpolatedimage.py:1185]: Could you explain the `+ 1.0` here?

File: jax_galsim/interpolatedimage.py [hunk 5]
```diff
@@ -1247,6 +1235,10 @@ def _find_maxk(kim, max_maxk, thresh):
     # maxk from the image (computed by _inner_comp_find_maxk)
     # by max_maxk from above
     return jnp.minimum(
-        _inner_comp_find_maxk(kim.array, thresh, kx, ky),
+        # jax-galsim tends to be less conservative for maxk
+        # since compared to galsim, it does NOT require 5 rows
+        # of pixels in a row below the threshold.
+        # thus we add some pixels here to ensure the galsim tests pass.
```
Comment (ismael-mendoza) [inline jax_galsim/interpolatedimage.py]: you only add one in the end no? 
```suggestion
        # thus we add one pixel here to ensure the galsim tests pass.
```

File: tests/jax/test_interpolatedimage_utils.py [hunk 6]
```diff
@@ -233,8 +233,9 @@ def test_interpolatedimage_utils_comp_to_galsim(
         x_interpolant=x_interp,
     )
 
-    np.testing.assert_allclose(gii.stepk, jgii.stepk, rtol=0.5, atol=0)
-    np.testing.assert_allclose(gii.maxk, jgii.maxk, rtol=0.5, atol=0)
+    np.testing.assert_allclose(jgii.stepk, gii.stepk, rtol=0, atol=1e-6)
+    np.testing.assert_allclose(jgii.maxk, gii.maxk, rtol=0.5, atol=0)
```
Comment (ismael-mendoza) [inline tests/jax/test_interpolatedimage_utils.py:238]:
```suggestion
    #FIXME: match maxk
    np.testing.assert_allclose(jgii.maxk, gii.maxk, rtol=0.5, atol=0)
```

**Review by ismael-mendoza — COMMENTED** [review]
Just a few questions and some changes. If you think the first two are actually a bug, not sure if these should go into a different PR...?

File: jax_galsim/interpolatedimage.py (hunk 5 above)
Comment (beckermr) [reply jax_galsim/interpolatedimage.py]: yes but I just guess and checked on what worked OK. there is no special reason one is the right answer.

File: jax_galsim/interpolatedimage.py
Comment (beckermr) [reply jax_galsim/interpolatedimage.py]: Yes good catch. I have added that.

Comment (beckermr) [reply jax_galsim/interpolatedimage.py]: We only need the minor for the force_maxk branch with the switch to the method above. Good question.

File: jax_galsim/interpolatedimage.py (hunk 4 above)
Comment (beckermr) [reply jax_galsim/interpolatedimage.py:1185]: done!

File: jax_galsim/interpolatedimage.py (hunk 5 above)
Comment (beckermr) [reply jax_galsim/interpolatedimage.py]: added a comment.

File: jax_galsim/interpolatedimage.py (hunk 2 above)
Comment (beckermr) [reply jax_galsim/interpolatedimage.py]: I am pretty sure this code branch was actually never executed due to the parent class having the same branch and thus skipping this call. I am double checking now.

**Comment by beckermr** [comment]
Let's see if removing this unused code works OK. I think it might and I do not recall why or what I was trying to do before.

**Comment by beckermr** [comment]
OK @ismael-mendoza! Thank you for the very helpful code review! This is ready for another look.

File: jax_galsim/interpolatedimage.py [hunk 7]
```diff
@@ -222,14 +219,19 @@ def _maxk(self):
         if self._jax_aux_data["_force_maxk"] > 0:
             return self._jax_aux_data["_force_maxk"]
         else:
-            return super()._maxk
+            # galsim uses a different way to handfle the WCS effects on maxk
```
Comment (ismael-mendoza) [inline jax_galsim/interpolatedimage.py]:
```suggestion
            # galsim uses a different way to handle the WCS effects on maxk
```

**Comment by beckermr** [comment]
@ismael-mendoza I recalled the general reason for the two classes here and added some notes on it.

**Review by ismael-mendoza — APPROVED** [review]
Thanks LGTM!
